### PR TITLE
Dynesty: add nested samples and write to backup file

### PR DESCRIPTION
**Pull request by sum33it**
This PR addresses two issues:
i) Added an option to write nested samples in file.
ii) Fix the issue of 'not writing into backup file'.

### CODE REVIEW

**Comment by ahnitz** [comment]
@sum33it Please address the CC issues.

**Comment by cdcapano** [comment]
@sum33it @ahnitz Is there any reason not to store the nested samples in the `samples` group? That would have symmetry with the MCMC samplers, since the samples written there for them are not really the posterior either. If you store the `logwt`, you could then load the posterior from the file on the file, or, using the extra args parser, tell it to just load the nested samples. So, under this model, the nested sampler doesn't even try to write the posterior at the end, it just writes all the nested samples. You then use extract samples to generate the posterior file, just like we do with MCMCs.

**Comment by ahnitz** [comment]
@sum33it What did you think of @cdcapano's suggestion?

**Comment by ahnitz** [comment]
@sum33it Poke

File: pycbc/inference/io/dynesty.py [hunk 1]
```diff
@@ -25,12 +25,96 @@
...
+                             "a parser is provided.")
+        if skip_args is None:
+            skip_args = []
+        actions = []
+
+        if 'raw_samples' not in skip_args:
+            act = parser.add_argument(
+                "--raw-samples", type=numpy.array, default=False,
+                help="Raw samples are the unweighted samples obtained from "
+                     "dynesty sampler."
```
Comment (cdcapano) [inline pycbc/inference/io/dynesty.py:73]: "dynesty sampler" -> "nested sampler"

File: pycbc/inference/io/dynesty.py [hunk 2]
```diff
@@ -25,12 +25,96 @@
...
+        elif kwargs:
+            raise ValueError("No other keyword arguments should be provded if "
+                             "a parser is provided.")
+        if skip_args is None:
+            skip_args = []
+        actions = []
+
+        if 'raw_samples' not in skip_args:
+            act = parser.add_argument(
+                "--raw-samples", type=numpy.array, default=False,
```
Comment (cdcapano) [inline pycbc/inference/io/dynesty.py:71]: `action='store_true'`

Comment (cdcapano) [reply pycbc/inference/io/dynesty.py:71]: And remove `type=numpy.array`

File: pycbc/inference/io/dynesty.py (hunk 1 above)
Comment (sum33it) [reply pycbc/inference/io/dynesty.py:75]: But this particular function deals specifically with dynesty sampler.

**Comment by sum33it** [comment]
@cdcapano  I guess it is in good shape. I have done the required tests and it works fine:
i) Writing raw samples to inference output file along with log-weights.
ii) `pycbc_inference_extract_sanples` and `pycbc_inference_plot_posterior' executables tested.

**Comment by ahnitz** [comment]
@cdcapano Is there any reason this cannot now be merged?

File: pycbc/inference/io/dynesty.py (hunk 1 above)
Comment (cdcapano) [reply pycbc/inference/io/dynesty.py:75]: Ok, fair enough

**Review by cdcapano — APPROVED** [review]
Looks good!

**Pull request merged**